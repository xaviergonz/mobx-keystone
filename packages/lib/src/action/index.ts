export { ActionCall, applyAction } from "./applyAction"
export {
  ActionContext,
  ActionContextActionType,
  ActionContextAsyncStepType,
  getCurrentActionContext,
} from "./context"
export { ActionMiddleware, ActionMiddlewareDisposer, addActionMiddleware } from "./middleware"
export { isModelAction, modelAction } from "./modelAction"
export { FlowRet, isModelFlow, modelFlow } from "./modelFlow"
export { runUnprotected } from "./protection"
export { isHookAction, isSpecialAction, SpecialAction } from "./specialActions"
