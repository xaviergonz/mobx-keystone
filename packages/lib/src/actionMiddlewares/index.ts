export {
  ActionTrackingMiddleware,
  actionTrackingMiddleware,
  ActionTrackingResult,
  ActionTrackingReturn,
  SimpleActionContext,
  simplifyActionContext,
} from "./actionTrackingMiddleware"
export {
  deserializeActionCall,
  deserializeActionCallArgument,
  onActionMiddleware,
  serializeActionCall,
  serializeActionCallArgument,
} from "./onActionMiddleware"
export { readonlyMiddleware, ReadonlyMiddlewareReturn } from "./readonlyMiddleware"
export { transaction, transactionMiddleware } from "./transactionMiddleware"
export { UndoEvent, UndoManager, undoMiddleware, UndoStore, withoutUndo } from "./undoMiddleware"
