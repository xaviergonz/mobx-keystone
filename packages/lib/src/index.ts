export * from "./action"
export * from "./actionMiddlewares"
export * from "./frozen"
export * from "./model"
export * from "./parent"
export * from "./patch"
export * from "./redux"
export * from "./ref"
export * from "./rootStore"
export * from "./snapshot"
export * from "./tweaker"
export { MobxDataModelError } from "./utils"
