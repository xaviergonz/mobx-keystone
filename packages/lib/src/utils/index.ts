import * as mobx from "mobx"
import {
  IObservableArray,
  isObservableArray,
  isObservableMap,
  isObservableObject,
  isObservableSet,
  ObservableMap,
  ObservableSet,
} from "mobx"
import { NonEmptyArray, PrimitiveValue } from "./types"

/**
 * A mobx-keystone error.
 */
export class MobxKeystoneError extends Error {
  constructor(msg: string) {
    super(msg)

    // Set the prototype explicitly.
    Object.setPrototypeOf(this, MobxKeystoneError.prototype)
  }
}

/**
 * @ignore
 * @internal
 */
export function failure(msg: string) {
  return new MobxKeystoneError(msg)
}

const writableHiddenPropDescriptor: PropertyDescriptor = {
  enumerable: false,
  writable: true,
  configurable: false,
  value: undefined,
}

/**
 * @ignore
 * @internal
 */
export function addHiddenProp(object: any, propName: PropertyKey, value: any, writable = true) {
  if (writable) {
    Object.defineProperty(object, propName, writableHiddenPropDescriptor)
    object[propName] = value
  } else {
    Object.defineProperty(object, propName, {
      enumerable: false,
      writable,
      configurable: true,
      value,
    })
  }
}

/**
 * @ignore
 * @internal
 */
export function makePropReadonly<T>(object: T, propName: keyof T, enumerable: boolean) {
  const propDesc = Object.getOwnPropertyDescriptor(object, propName)
  if (propDesc) {
    propDesc.enumerable = enumerable
    if (propDesc.get) {
      delete propDesc.set
    } else {
      propDesc.writable = false
    }
    Object.defineProperty(object, propName, propDesc)
  }
}

/**
 * @ignore
 * @internal
 */
export function isPlainObject(value: any): value is Object {
  if (!isObject(value)) return false
  const proto = Object.getPrototypeOf(value)
  return proto === Object.prototype || proto === null
}

/**
 * @ignore
 * @internal
 */
export function isObject(value: any): value is Object {
  return value !== null && typeof value === "object"
}

/**
 * @ignore
 * @internal
 */
export function isPrimitive(value: any): value is PrimitiveValue {
  switch (typeof value) {
    case "number":
    case "string":
    case "boolean":
    case "undefined":
    case "bigint":
      return true
  }
  return value === null
}

/**
 * @ignore
 * @internal
 */
export function debugFreeze(value: object) {
  if (inDevMode()) {
    Object.freeze(value)
  }
}

/**
 * @ignore
 * @internal
 */
export function deleteFromArray<T>(array: T[], value: T): boolean {
  let index = array.indexOf(value)
  if (index >= 0) {
    array.splice(index, 1)
    return true
  }
  return false
}

/**
 * @ignore
 * @internal
 */
export function isMap(val: any): val is Map<any, any> | ObservableMap {
  return val instanceof Map || isObservableMap(val)
}

/**
 * @ignore
 * @internal
 */
export function isSet(val: any): val is Set<any> | ObservableSet {
  return val instanceof Set || isObservableSet(val)
}

/**
 * @ignore
 * @internal
 */
export function isArray(val: any): val is any[] | IObservableArray {
  return Array.isArray(val) || isObservableArray(val)
}

/**
 * Checks whether an array is non-empty.
 *
 * @param arr Array to check.
 * @returns `true` if `arr` is non-empty, otherwise `false`.
 */
export function isNonEmptyArray<T>(arr: T[]): arr is NonEmptyArray<T> {
  return arr.length !== 0
}

/**
 * @ignore
 * @internal
 */
export function inDevMode(): boolean {
  return process.env.NODE_ENV !== "production"
}

/**
 * @ignore
 * @internal
 */
export function assertIsObject(value: unknown, argName: string): asserts value is object {
  if (!isObject(value)) {
    throw failure(`${argName} must be an object`)
  }
}

/**
 * @ignore
 * @internal
 */
export function assertIsPlainObject(value: unknown, argName: string): asserts value is object {
  if (!isPlainObject(value)) {
    throw failure(`${argName} must be a plain object`)
  }
}

/**
 * @ignore
 * @internal
 */
export function assertIsObservableObject(value: unknown, argName: string): asserts value is object {
  if (!isObservableObject(value)) {
    throw failure(`${argName} must be an observable object`)
  }
}

/**
 * @ignore
 * @internal
 */
export function assertIsObservableArray(
  value: unknown,
  argName: string
): asserts value is IObservableArray {
  if (!isObservableArray(value)) {
    throw failure(`${argName} must be an observable array`)
  }
}

/**
 * @ignore
 * @internal
 */
export function assertIsMap(value: unknown, argName: string): asserts value is Map<any, any> {
  if (!isMap(value)) {
    throw failure(`${argName} must be a map`)
  }
}

/**
 * @ignore
 * @internal
 */
export function assertIsSet(value: unknown, argName: string): asserts value is Set<any> {
  if (!isSet(value)) {
    throw failure(`${argName} must be a set`)
  }
}

/**
 * @ignore
 * @internal
 */
export function assertIsFunction(value: unknown, argName: string): asserts value is Function {
  if (typeof value !== "function") {
    throw failure(`${argName} must be a function`)
  }
}

/**
 * @ignore
 * @internal
 */
export function assertIsPrimitive(
  value: unknown,
  argName: string
): asserts value is PrimitiveValue {
  if (!isPrimitive(value)) {
    throw failure(`${argName} must be a primitive`)
  }
}

/**
 * @ignore
 * @internal
 */
export function assertIsString(value: unknown, argName: string): asserts value is string {
  if (typeof value !== "string") {
    throw failure(`${argName} must be a string`)
  }
}

/**
 * @ignore
 * @internal
 */
export interface DecorateMethodOrFieldData {
  target: any
  propertyKey: string
  baseDescriptor?: PropertyDescriptor & { initializer?: () => any }
}

const decoratorsSymbol = Symbol("decorators")

type WrapFunction = (data: DecorateMethodOrFieldData, fn: any) => any
type LateInitializationFunctionsArray = ((instance: any) => void)[]

/**
 * @ignore
 * @internal
 */
export function addLateInitializationFunction(target: any, fn: (instance: any) => void) {
  let decoratorsArray: LateInitializationFunctionsArray = target[decoratorsSymbol]
  if (!decoratorsArray) {
    decoratorsArray = []
    addHiddenProp(target, decoratorsSymbol, decoratorsArray)
  }
  decoratorsArray.push(fn)
}

/**
 * @ignore
 * @internal
 */
export function decorateWrapMethodOrField(
  decoratorName: string,
  data: DecorateMethodOrFieldData,
  wrap: WrapFunction
): any {
  const { target, propertyKey, baseDescriptor } = data

  const addFieldDecorator = () => {
    addLateInitializationFunction(target, (instance) => {
      instance[propertyKey] = wrap(data, instance[propertyKey])
    })
  }

  if (baseDescriptor) {
    if (baseDescriptor.get !== undefined) {
      throw failure(`@${decoratorName} cannot be used with getters`)
    }

    if (baseDescriptor.value) {
      // babel / typescript - method decorator
      // @action method() { }
      return {
        enumerable: false,
        writable: true,
        configurable: true,
        value: wrap(data, baseDescriptor.value),
      }
    } else {
      // babel - field decorator: @action method = () => {}
      addFieldDecorator()
    }
  } else {
    // typescript - field decorator
    addFieldDecorator()
  }
}

/**
 * @ignore
 * @internal
 */
export function runLateInitializationFunctions(instance: any): void {
  const fns: LateInitializationFunctionsArray | undefined = instance[decoratorsSymbol]
  if (fns) {
    for (const fn of fns) {
      fn(instance)
    }
  }
}

const warningsAlreadyDisplayed = new Set<string>()

/**
 * @ignore
 * @internal
 */
export function logWarning(type: "warn" | "error", msg: string, uniqueKey?: string): void {
  if (uniqueKey) {
    if (warningsAlreadyDisplayed.has(uniqueKey)) {
      return
    }
    warningsAlreadyDisplayed.add(uniqueKey)
  }

  msg = "[mobx-keystone] " + msg
  switch (type) {
    case "warn":
      console.warn(msg)
      break
    case "error":
      console.error(msg)
      break
    default:
      throw failure(`unknown log type - ${type}`)
  }
}

const notMemoized = Symbol("notMemoized")

/**
 * @ignore
 * @internal
 */
export function lateVal<TF extends (...args: any[]) => any>(getter: TF): TF {
  let memoized: TF | typeof notMemoized = notMemoized

  const fn = (...args: any[]): any => {
    if (memoized === notMemoized) {
      memoized = getter(...args)
    }
    return memoized
  }

  return fn as TF
}

/**
 * @ignore
 * @internal
 */
export function lazy<V>(valueGen: () => V): () => V {
  let inited = false
  let val: V | undefined

  return (): V => {
    if (!inited) {
      val = valueGen()
      inited = true
    }
    return val!
  }
}

// just to ensure import * is kept properly
/**
 * @ignore
 * @internal
 */
export const mobx6 = {
  // eslint-disable-next-line no-useless-concat
  makeObservable: (mobx as any)["makeObservable" + ""] as typeof mobx["makeObservable"],
}

/**
 * @ignore
 * @internal
 */
export function getMobxVersion(): number {
  if (mobx6.makeObservable!) {
    return 6
  } else {
    return 5
  }
}
