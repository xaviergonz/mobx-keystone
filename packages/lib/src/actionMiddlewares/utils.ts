import { modelIdKey } from "../model/metadata"
import { isModel } from "../model/utils"
import type { RootPath } from "../parent/path"
import type { Path } from "../parent/pathTypes"

/**
 * @internal
 */
export function rootPathToTargetPathIds(rootPath: RootPath<any>): (string | null)[] {
  const targetPathIds: (string | null)[] = []

  for (let i = 0; i < rootPath.path.length; i++) {
    const targetObj = rootPath.pathObjects[i + 1] // first is root, we don't care about its ID
    const targetObjId = isModel(targetObj) ? targetObj[modelIdKey] ?? null : null
    targetPathIds.push(targetObjId)
  }

  return targetPathIds
}

/**
 * @internal
 */
export function pathToTargetPathIds(root: any, path: Path): (string | null)[] {
  const targetPathIds: (string | null)[] = []
  let current = root // we don't care about the root ID

  for (let i = 0; i < path.length; i++) {
    current = current[path[i]]
    const targetObjId = isModel(current) ? current[modelIdKey] ?? null : null
    targetPathIds.push(targetObjId)
  }

  return targetPathIds
}
