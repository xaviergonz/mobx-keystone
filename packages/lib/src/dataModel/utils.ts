import type { ModelClass } from "../modelShared/BaseModelShared"
import { failure } from "../utils"
import type { AnyDataModel } from "./BaseDataModel"
import { _BaseDataModel } from "./_BaseDataModel"

/**
 * Checks if an object is a data model instance.
 *
 * @param model
 * @returns
 */
export function isDataModel(model: unknown): model is AnyDataModel {
  return model instanceof _BaseDataModel
}

/**
 * @ignore
 * @internal
 *
 * Asserts something is actually a data model.
 *
 * @param model
 * @param argName
 */
export function assertIsDataModel(
  model: unknown,
  argName: string,
  customErrMsg = "must be a data model instance"
): asserts model is AnyDataModel {
  if (!isDataModel(model)) {
    throw failure(`${argName} ${customErrMsg}`)
  }
}

/**
 * @ignore
 * @internal
 */
export function isDataModelClass(modelClass: unknown): modelClass is ModelClass<AnyDataModel> {
  if (typeof modelClass !== "function") {
    return false
  }

  if (modelClass !== _BaseDataModel && !(modelClass.prototype instanceof _BaseDataModel)) {
    return false
  }

  return true
}

/**
 * @ignore
 * @internal
 */
export function assertIsDataModelClass(
  modelClass: unknown,
  argName: string
): asserts modelClass is ModelClass<AnyDataModel> {
  if (typeof modelClass !== "function") {
    throw failure(`${argName} must be a class`)
  }

  if (modelClass !== _BaseDataModel && !(modelClass.prototype instanceof _BaseDataModel)) {
    throw failure(`${argName} must extend DataModel`)
  }
}
