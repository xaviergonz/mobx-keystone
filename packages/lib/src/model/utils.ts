import { SnapshotInOfModel } from "../snapshot/SnapshotOf"
import { failure, isPlainObject } from "../utils"
import { modelMetadataKey } from "./metadata"
import { AnyModel, BaseModel, ModelClass } from "./Model"

/**
 * Checks if an object is a model instance.
 *
 * @param model
 * @returns
 */
export function isModel(model: any): model is AnyModel {
  return model instanceof BaseModel
}

/**
 * @ignore
 *
 * Asserts something is actually a model.
 *
 * @param model
 * @param argName
 */
export function assertIsModel(
  model: AnyModel,
  argName: string,
  customErrMsg = "must be a model instance"
) {
  if (!isModel(model)) {
    throw failure(`${argName} ${customErrMsg}`)
  }
}

/**
 * @ignore
 */
export function isModelClass(modelClass: any): modelClass is ModelClass<AnyModel> {
  if (typeof modelClass !== "function") {
    return false
  }

  if (modelClass !== BaseModel && !(modelClass.prototype instanceof BaseModel)) {
    return false
  }

  return true
}

/**
 * @ignore
 */
export function assertIsModelClass(modelClass: ModelClass<AnyModel>, argName: string): void {
  if (typeof modelClass !== "function") {
    throw failure(`${argName} must be a class`)
  }

  if (modelClass !== BaseModel && !(modelClass.prototype instanceof BaseModel)) {
    throw failure(`${argName} must extend Model`)
  }
}

/**
 * @ignore
 */
export function isModelSnapshot(sn: any): sn is SnapshotInOfModel<AnyModel> {
  return isPlainObject(sn) && !!sn[modelMetadataKey]
}

/**
 * @ignore
 */
export function checkModelDecoratorArgs(fnName: string, target: any, propertyKey: string) {
  if (typeof propertyKey !== "string") {
    throw failure(fnName + " cannot be used over symbol properties")
  }

  const errMessage = fnName + " must be used over model classes or instances"

  if (!target) {
    throw failure(errMessage)
  }

  // check target is a model object or extended class
  if (
    !(target instanceof BaseModel) &&
    target !== BaseModel &&
    !(target.prototype instanceof BaseModel)
  ) {
    throw failure(errMessage)
  }
}
