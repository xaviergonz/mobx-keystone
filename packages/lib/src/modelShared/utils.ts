import type { AnyDataModel } from "../dataModel/BaseDataModel"
import { isDataModelClass } from "../dataModel/utils"
import type { AnyModel } from "../model/BaseModel"
import { isModelClass } from "../model/utils"
import { failure } from "../utils"
import type { ModelClass } from "./BaseModelShared"

/**
 * @internal
 *
 * Asserts something is actually a class or data model.
 *
 * @param model
 * @param argName
 */
export function assertIsClassOrDataModelClass(
  model: unknown,
  argName: string,
  customErrMsg = "must be a class or data model class"
): asserts model is ModelClass<AnyModel> | ModelClass<AnyDataModel> {
  if (!isModelClass(model) && !isDataModelClass(model)) {
    throw failure(`${argName} ${customErrMsg}`)
  }
}
