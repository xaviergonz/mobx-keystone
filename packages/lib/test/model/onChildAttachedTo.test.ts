import {
  BaseModel,
  getParentToChildPath,
  Model,
  model,
  modelAction,
  onChildAttachedTo,
  Path,
  prop,
} from "../../src"
import { isArray, isObject } from "../../src/utils"
import "../commonSetup"

const name = (obj: object) => {
  if (obj instanceof BaseModel) return obj.$modelType
  else if (isArray(obj)) return "Array"
  else if (isObject(obj)) return "Object"
  else return "unknown"
}

const events: string[] = []

const log = (type: "attached" | "detached", self: object, child: object, path: Path) => {
  events.push(
    `${type} ${type === "attached" ? "to" : "from"} ${name(self)} <- ${name(child)} at [${path.join(
      "/"
    )}]`
  )
}

beforeEach(() => {
  events.length = 0
})
;[true, false].forEach((deep) => {
  test(`onChildAttachedTo (deep: ${deep})`, () => {
    const disposers: ((runDetach: boolean) => void)[] = []

    const addHook = (node: object) => {
      const d = onChildAttachedTo(
        () => node,
        (child) => {
          const path = getParentToChildPath(node, child)!
          if (!path) {
            fail("path between " + node + " and " + child + " could not be found")
          }
          log("attached", node, child, path)
          return () => {
            log("detached", node, child, path)
          }
        },
        {
          fireForCurrentChildren: true,
          deep: deep,
        }
      )
      disposers.push(d)
    }

    @model(`R-deep:${deep}`)
    class R extends Model({
      a: prop<A>(() => new A({})),
    }) {
      onInit() {
        addHook(this)
      }

      @modelAction
      setA(a: A) {
        this.a = a
      }
    }

    @model(`A-deep:${deep}`)
    class A extends Model({
      b: prop<B>(() => new B({})),
    }) {
      onInit() {
        addHook(this)
      }

      @modelAction
      setB(b: B) {
        this.b = b
      }
    }

    @model(`B-deep:${deep}`)
    class B extends Model({
      primitive: prop<string>("hi"),
      arr: prop<number[]>(() => [1, 2, 3]),
    }) {
      onInit() {
        addHook(this)
      }

      @modelAction
      setArr(arr: number[]) {
        this.arr = arr
      }
    }

    expect(events).toHaveLength(0)

    // initial children should be attached
    const r = new R({})
    expect(events).toMatchSnapshot("initial")

    // setting a new A
    events.length = 0
    r.setA(new A({}))
    expect(events).toMatchSnapshot("new A")

    // setting a new B
    events.length = 0
    r.a.setB(new B({}))
    expect(events).toMatchSnapshot("new B")

    // setting a new arr
    events.length = 0
    r.a.b.setArr([4, 5, 6])
    expect(events).toMatchSnapshot("new arr")

    // disposer
    disposers.forEach((d) => d(false))
    events.length = 0
    r.a.b.setArr([4, 5, 6])
    expect(events).toHaveLength(0)
  })
})

test("dynamic target", () => {
  const disposers: ((runDetach: boolean) => void)[] = []

  const addHook = (fn: () => object) => {
    const d = onChildAttachedTo(
      fn,
      (child) => {
        const node = fn()
        const path = getParentToChildPath(node, child)!
        log("attached", node, child, path)
        return () => {
          log("detached", node, child, path)
        }
      },
      {
        fireForCurrentChildren: true,
        deep: false,
      }
    )
    disposers.push(d)
  }

  @model(`TodoList`)
  class TodoList extends Model({
    todos: prop<Todo[]>(() => []),
  }) {
    onInit() {
      addHook(() => this.todos)
    }

    @modelAction
    addTodo(todo: Todo) {
      this.todos = [...this.todos, todo]
    }

    @modelAction
    removeTodo(todo: Todo) {
      this.todos = this.todos.filter((t) => t !== todo)
    }
  }

  @model(`Todo`)
  class Todo extends Model({}) {}

  expect(events).toHaveLength(0)

  // initial children should be attached
  const todoList = new TodoList({
    todos: [new Todo({}), new Todo({})],
  })
  expect(events).toMatchSnapshot("initial")

  // add a new todo
  events.length = 0
  todoList.addTodo(new Todo({}))
  expect(events).toMatchSnapshot("add Todo")

  // remove todo
  events.length = 0
  todoList.removeTodo(todoList.todos[1])
  expect(events).toMatchSnapshot("remove Todo")

  // disposer
  disposers.forEach((d) => d(false))
  events.length = 0
  todoList.removeTodo(todoList.todos[0])
  expect(events).toHaveLength(0)
})
