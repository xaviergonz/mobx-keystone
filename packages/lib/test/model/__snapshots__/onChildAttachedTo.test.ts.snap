// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dynamic target: add Todo 1`] = `
Array [
  "attached to Array <- Todo at [2]",
]
`;

exports[`dynamic target: initial 1`] = `
Array [
  "attached to Array <- Todo at [0]",
  "attached to Array <- Todo at [1]",
]
`;

exports[`dynamic target: remove Todo 1`] = `
Array [
  "detached from Array <- Todo at [1]",
]
`;

exports[`onChildAttachedTo (deep: false): initial 1`] = `
Array [
  "attached to B-deep:false <- Array at [$/arr]",
  "attached to A-deep:false <- B-deep:false at [$/b]",
  "attached to R-deep:false <- A-deep:false at [$/a]",
]
`;

exports[`onChildAttachedTo (deep: false): new A 1`] = `
Array [
  "attached to B-deep:false <- Array at [$/arr]",
  "attached to A-deep:false <- B-deep:false at [$/b]",
  "detached from R-deep:false <- A-deep:false at [$/a]",
  "attached to R-deep:false <- A-deep:false at [$/a]",
]
`;

exports[`onChildAttachedTo (deep: false): new B 1`] = `
Array [
  "attached to B-deep:false <- Array at [$/arr]",
  "detached from A-deep:false <- B-deep:false at [$/b]",
  "attached to A-deep:false <- B-deep:false at [$/b]",
]
`;

exports[`onChildAttachedTo (deep: false): new arr 1`] = `
Array [
  "detached from B-deep:false <- Array at [$/arr]",
  "attached to B-deep:false <- Array at [$/arr]",
]
`;

exports[`onChildAttachedTo (deep: true): initial 1`] = `
Array [
  "attached to B-deep:true <- Array at [$/arr]",
  "attached to A-deep:true <- B-deep:true at [$/b]",
  "attached to A-deep:true <- Array at [$/b/$/arr]",
  "attached to R-deep:true <- A-deep:true at [$/a]",
  "attached to R-deep:true <- B-deep:true at [$/a/$/b]",
  "attached to R-deep:true <- Array at [$/a/$/b/$/arr]",
]
`;

exports[`onChildAttachedTo (deep: true): new A 1`] = `
Array [
  "attached to B-deep:true <- Array at [$/arr]",
  "attached to A-deep:true <- B-deep:true at [$/b]",
  "attached to A-deep:true <- Array at [$/b/$/arr]",
  "detached from R-deep:true <- Array at [$/a/$/b/$/arr]",
  "detached from R-deep:true <- B-deep:true at [$/a/$/b]",
  "detached from R-deep:true <- A-deep:true at [$/a]",
  "attached to R-deep:true <- A-deep:true at [$/a]",
  "attached to R-deep:true <- B-deep:true at [$/a/$/b]",
  "attached to R-deep:true <- Array at [$/a/$/b/$/arr]",
]
`;

exports[`onChildAttachedTo (deep: true): new B 1`] = `
Array [
  "attached to B-deep:true <- Array at [$/arr]",
  "detached from A-deep:true <- Array at [$/b/$/arr]",
  "detached from A-deep:true <- B-deep:true at [$/b]",
  "attached to A-deep:true <- B-deep:true at [$/b]",
  "attached to A-deep:true <- Array at [$/b/$/arr]",
  "detached from R-deep:true <- Array at [$/a/$/b/$/arr]",
  "detached from R-deep:true <- B-deep:true at [$/a/$/b]",
  "attached to R-deep:true <- B-deep:true at [$/a/$/b]",
  "attached to R-deep:true <- Array at [$/a/$/b/$/arr]",
]
`;

exports[`onChildAttachedTo (deep: true): new arr 1`] = `
Array [
  "detached from B-deep:true <- Array at [$/arr]",
  "attached to B-deep:true <- Array at [$/arr]",
  "detached from A-deep:true <- Array at [$/b/$/arr]",
  "attached to A-deep:true <- Array at [$/b/$/arr]",
  "detached from R-deep:true <- Array at [$/a/$/b/$/arr]",
  "attached to R-deep:true <- Array at [$/a/$/b/$/arr]",
]
`;
